library/core/src/main/java/com/google/android/exoplayer2/source/SampleQueue.java

```diff
@@ -767,7 +767,8 @@ private synchronized boolean attemptSplice(long timeUs) {
 
   private long discardUpstreamSampleMetadata(int discardFromIndex) {
     int discardCount = getWriteIndex() - discardFromIndex;
-    Assertions.checkArgument(0 <= discardCount && discardCount <= (length - readPosition));
+    Assertions.checkArgument(0 <= discardCount);
+    discardCount = getMaximumDiscardCount(discardCount);
```

Comment:
Could this cause the `java.lang.IndexOutOfBoundsException` with `SampleDataQueue` I saw in my testing?

BTW the ConcurrentModificationException from my pull request.  @tvarga-dss feel free to collapse out any of my commits in your re-base to simplicity this, I just want to see the future supported ASAP.  If you are unable to work on this please let us know.

Reply:
> Could this cause the java.lang.IndexOutOfBoundsException with SampleDataQueue I saw in my testing?

Sounds like a likely reason, yes.

> BTW the ConcurrentModificationException from my pull request.

I added some explanations further down why I think access to the first sample indices is safe. But I only now realized this is not the concurrent modification you were talking about. :)  I'll check against your comments made in https://github.com/google/ExoPlayer/pull/6321#issuecomment-570344663 to see if this can still happen.